File: pkg/__test__/proto_test.ts
//@ts-ignore
import { encode } from "https://deno.land/x/msgpack@v1.2/mod.ts";

const Query = {
  "db": "database.sqlite",
  "commands": [
    {"method": "In", "args": ["data"]},
    {
        "method": "Find",
        "args": [{
            "a": 1,
            "b": {
                "$gt": 2,
            },
        }]
    },
  ],
};

const encoded: Uint8Array = encode(Query);
//@ts-ignore
Deno.writeFileSync("proto_test.msgpack", encoded);

File: pkg/__test__/proto_test.go
package tests

import (
	"fmt"
	"os"
	"testing"

	"pkg/adapter"
	"pkg/proto"

	_ "github.com/mattn/go-sqlite3"
)

func TestProtoMessagePack(t *testing.T) {
	message, err := os.ReadFile("proto_test.msgpack")
	if err != nil {
		t.Fatalf("failed to read file: %v", err)
	}
	req := proto.Request{}
	fmt.Println(message)
	req.UnmarshalMsg(message)
	query, err := req.Parse(adapter.CommonDialect{})
	if err != nil {
		t.Fatalf("failed to parse query: %v", err)
	}
	db := "database.sqlite"
	if query.Db != db {
		t.Fatalf("expected db %s, got %s", db, query.Db)
	}
	expr := "SELECT * FROM data WHERE a = ? AND b > ?"
	if query.Expression != expr {
		t.Fatalf("expected expression %s, got %s", expr, query.Expression)
	}
	//fmt.Println(q)
}
